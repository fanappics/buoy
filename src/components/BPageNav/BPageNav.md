### Example Usage

```js
let links = [
  { label: 'Lone Required Input', href: 'bpn-input-1', completed: false },
  { label: 'Email', href: 'bpn-input-2', completed: false }
]

<div style="display: flex">
  <div style="margin-right: 1rem;">
    <b-page-nav :links="links" column />
  </div>
  <div style="flex-grow: 1">
    <b-validator scope="bpn-scope-1" v-model="links[0].completed">
      <b-input id='bpn-input-1' label='Lone Required Input' required />
    </b-validator>
    <br>
    <b-container label="Header">
      <b-validator scope="bpn-scope-2" v-model="links[1].completed">
        <b-input id="bpn-input-2" type="email" label="Email" required />
      </b-validator>
    </b-container>
  </div>
</div>
```

### Column Example With Accordions

```js
let links = [
  { label: 'Lone Input', href: 'bpn-input-3', completed: true },
  { label: 'Header', href: 'bpn-header-1', completed: false, event: 'expand-bpn-header-1' },
  { label: 'Collapsed Header', href: 'bpn-header-2', completed: false, event: 'expand-bpn-header-2' }
]

<div style="display: flex">
  <div style="margin-right: 1rem;">
    <b-page-nav :links="links" column />
  </div>
  <div style="flex-grow: 1">
    <b-input id='bpn-input-3' label='Lone Input' />
    <br>
    <b-accordion id="bpn-header-1" label="Header">
      <b-validator scope="bpn-scope-3" v-model="links[1].completed">
        <b-input id="bpn-input-4" type="email" label="Email" required />
      </b-validator>
    </b-accordion>
    <b-accordion id="bpn-header-2" label="Collapsed Header" collapsed>
      <b-validator scope="bpn-scope-4" v-model="links[2].completed">
        <b-textarea id="bpn-textarea-1" label="Write me a novel" required />
      </b-validator>
    </b-accordion>
  </div>
</div>
```
