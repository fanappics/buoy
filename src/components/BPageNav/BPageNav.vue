<template>
  <nav :id="id">
      <ul role="menu" aria-label="Page Navigation">
        <li v-for="(link, index) in links" role ="menuitem" :key="index" :class="{ 'inline': !column }">
          <a
            :href="`#${link.href}`"
            :aria-label="`${link.label} ${link.completed ? 'Completed' : 'Uncompleted'}`"
            :class="{ 'inline': !column }"
            @click.prevent.stop="jumpTo(link)"
            @keyup.enter.prevent="jumpTo(link)"
            @keydown.enter.prevent
            @keyup.up="upHandler($event)"
            @keydown.up.prevent
            @keyup.left="upHandler($event)"
            @keydown.left.prevent
            @keyup.right="downHandler($event)"
            @keydown.right.prevent
            @keyup.down="downHandler($event)"
            @keydown.down.prevent
          >
            <i v-if="link.completed" class="icon-complete ion-ios-checkmark" aria-hidden="true" />
            <i v-else class="icon-incomplete ion-record" aria-hidden="true" />
            {{ link.label }}
          </a>
        </li>
      </ul>
    </nav>
</template>

<script>
import events from '../../event-bus'

export default {
  name: 'b-page-nav',
  props: {
    //Required props
    /** 
    * <pre style="font-family: monospace;">
    * Array of links to be rendered
    * Each link is an object containing:
    * |-------------|-----------|------------|----------------------------------------------------------|
    * |  Prop name  |   Type    |  Default   |                      Description                         |
    * |-------------|-----------|------------|----------------------------------------------------------|
    * |  label      |  string   |  required  |  Display string of the link                              |
    * |  href       |  string   |  required  |  Id of the element to navigate to                        |
    * |  completed  |  boolean  |  optional  |  Specifies related section as ready for form submission  |
    * |  event      |  string   |  optional  |  Name of an event to emit on the event-bus on click      |
    * |-------------|-----------|------------|----------------------------------------------------------|
    * </pre>
    */
    links: {
      type: Array,
      required: true,
      validator (value) {
        for (let link of value) {
          if (!link.label || !link.href)
            throw new TypeError('Each link must have a label and href defined')
        }
        return true
      }
    },
    //Optional props
    /** 
    * Layout links in a column
    */
    column: Boolean,
    /** 
    * Override id on nav element
    */
    id: {
      type: String,
      default () {
        return `page-nav-${this._uid}`
      }
    }
  },
  data () {
    return {}
  },
  methods: {
    /**
    * @param {Object} link
    * Scrolls to and focuses the element pertaining to link.href
    * Also emits link.event if specified
    */
    jumpTo (link) {
      if (link.event) {
        events.$emit(link.event)
      }
      this.$nextTick(function () {
        const element = document.getElementById(link.href)
        if (element) {
          element.scrollIntoView()
          element.focus()
        }
      })
    },
    /**
     * Handles up and left arrow keyup events
     */
    upHandler (event) {
      const target = event.target.parentElement
      if(target.previousSibling)
        target.previousSibling.firstElementChild.focus()
    },
    /**
     * Handles right and down arrow keyup events
     */
    downHandler (event) {
      const target = event.target.parentElement
      if(target.nextSibling)
        target.nextSibling.firstElementChild.focus()
    }
  }
}
</script>

<style scoped>
  a {
    background-color: #F5F6F8;
    color: #333333;
    display: block;
    padding: .5rem .5rem .5rem 1rem;
    text-decoration: none;
  }

  a:focus, a:hover {
    color: #FFFFFF;
    background-color: #00AAED;
  }

  a:focus i, a:hover i {
    color: #FFFFFF;
  }

  hr {
    background-color: #DEDEDE;
    border: none;
    height: 10px;
    margin: 0;
    width: 1px;
  }

  i {
    font-size: 1rem;
    padding-right: .5rem;
  }

  ul {
    list-style: none;
    display: inline;
    margin: 0;
    padding: 0;
  }

  .flex {
    display: flex;
  }

  .icon-complete {
    color: #00AAED;
  }

  .icon-incomplete {
    color: #FFFFFF;
  }

  .inline {
    display: inline;
  }
</style>
