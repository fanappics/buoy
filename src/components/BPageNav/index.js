import BPageNav from './BPageNav.vue'

/* istanbul ignore next */
BPageNav.install = function install (Vue) {
  Vue.component(BPageNav.name, BPageNav)
}

export default BPageNav
