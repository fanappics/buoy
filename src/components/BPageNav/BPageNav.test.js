import { shallow } from 'vue-test-utils'
import BPageNav from './BPageNav'
import events from '../../event-bus'

describe('Page nav', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BPageNav, {
      propsData: {
        links: [{ label: 'Link', href: 'link-1', completed: true }]
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('fires optional event on click', () => {
    events.$emit = jest.fn()
    const wrapper = shallow(BPageNav, {
      propsData: {
        links: [{ label: 'Link', href: 'link-1', event: 'event' }]
      }
    })
    wrapper.find('nav > ul > li > a').trigger('keyup.enter')
    expect(events.$emit).toHaveBeenCalled()
  })
})

describe('Page nav multiple links', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BPageNav, {
      propsData: {
        links: [{ label: 'Link', href: 'link-1', completed: true }, { label: 'Link2', href: 'link-2', completed: false }]
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
