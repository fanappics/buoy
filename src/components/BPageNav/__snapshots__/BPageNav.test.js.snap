// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page nav mounted html should match snapshot 1`] = `
<nav id="page-nav-1">
    <ul role="menu" aria-label="Page Navigation">
        <li role="menuitem" class="inline">
            <a href="#link-1" aria-label="Link Completed" class="inline">
                <i aria-hidden="true" class="icon-complete ion-ios-checkmark"></i>
                Link
            </a>
        </li>
    </ul>
</nav>
`;

exports[`Page nav multiple links mounted html should match snapshot 1`] = `
<nav id="page-nav-2">
    <ul role="menu" aria-label="Page Navigation">
        <li role="menuitem" class="inline">
            <a href="#link-1" aria-label="Link Completed" class="inline">
                <i aria-hidden="true" class="icon-complete ion-ios-checkmark"></i>
                Link
            </a>
        </li>
        <li role="menuitem" class="inline">
            <a href="#link-2" aria-label="Link2 Uncompleted" class="inline">
                <i aria-hidden="true" class="icon-incomplete ion-record"></i>
                Link2
            </a>
        </li>
    </ul>
</nav>
`;
